In [1]:
import pandas as pd
import networkx as nx
from statistics import mean

In [2]:
node_feat = pd.read_csv('3437.feat', delim_whitespace=True, header=None)
node_feat.columns = ['Node'] + [f'feat_{i}' for i in range(1, node_feat.shape[1])]
node_feat

Node  feat_1  feat_2  feat_3  feat_4  feat_5  feat_6  feat_7  feat_8  \
0    3438       0       0       0       0       0       0       0       0   
1    3439       0       0       0       0       0       0       0       0   
2    3440       0       0       0       0       0       0       0       0   
3    3441       0       0       0       0       0       0       0       0   
4    3442       0       0       0       0       0       0       0       0   
..    ...     ...     ...     ...     ...     ...     ...     ...     ...   
145  3579       0       0       0       0       0       0       0       0   
146  3580       0       0       0       0       0       0       0       0   
147  3581       0       0       0       0       0       0       0       0   
148  3582       0       0       0       0       0       0       0       0   
149  3583       0       0       0       0       0       0       0       0   

     feat_9  ...  feat_253  feat_254  feat_255  feat_256  feat_257  feat_258  \
0         0  ...         0         0         0         0         0         0   
1         0  ...         0         0         0         0         0         0   
2         0  ...         0         0         0         0         0         0   
3         0  ...         0         0         0         0         0         0   
4         0  ...         0         0         0         0         0         0   
..      ...  ...       ...       ...       ...       ...       ...       ...   
145       0  ...         0         0         0         0         0         0   
146       0  ...         0         0         0         0         0         0   
147       0  ...         0         0         0         0         0         0   
148       0  ...         0         0         0         0         0         0   
149       0  ...         0         0         0         0         0         0   

     feat_259  feat_260  feat_261  feat_262  
0           0         0         0         0  
1           0         0         0         0  
2           0         0         0         0  
3           0         0         0         0  
4           0         0         0         0  
..        ...       ...       ...       ...  
145         0         0         0         0  
146         0         0         0         0  
147         0         0         0         0  
148         0         0         0         0  
149         0         0         0         0  

[150 rows x 263 columns]

In [3]:
edges = pd.read_csv('3437.edges', delim_whitespace=True, header=None,names=["node1","node2"])

edges

node1  node2
0     3562   3479
1     3517   3577
2     3510   3529
3     3545   3480
4     3492   3556
..     ...    ...
809   3512   3577
810   3521   3442
811   3508   3472
812   3469   3510
813   3456   1085

[814 rows x 2 columns]

In [4]:
G = nx.Graph() 

G.add_nodes_from(node_feat['Node'])

edge = [(row['node1'], row['node2']) for index, row in edges.iterrows()]
G.add_edges_from(edge)

In [8]:
degree = dict(G.degree())


In [9]:
print(f"Number of nodes: {G.number_of_nodes()}")
print(f"Number of edges: {G.number_of_edges()}")
print(f"Maximum degree: {max(degree.values())}")
print(f"Maximum degree: {mean(degree.values())}")

Number of nodes: 151
Number of edges: 407
Maximum degree: 25
Maximum degree: 5.3907284768211925


## We can see that user ID : 3545 has highest number of friend through this network 

In [34]:
target_value = max(degree.values())
keys_with_max_value = [key for key, value in degree.items() if value == target_value]

print( keys_with_max_value)

[3545]


## select 11 user among our network 

In [99]:
sequence = list(range(3438, 3449 ))

In [100]:
sequence

[3438, 3439, 3440, 3441, 3442, 3443, 3444, 3445, 3446, 3447, 3448]

In [101]:
subgraph=G.subgraph(sequence)

In [102]:
subgraph

In [103]:
subgraph.number_of_nodes()

11

In [104]:
subgraph.number_of_edges()

1

## After Applying Adj_matrix , we can say that users ID : 3448 and 3438 are friends in this small network 

In [105]:
adj_matrix = nx.to_numpy_array(subgraph)

In [106]:
adj_matrix

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

## the aim from Adj_matrix to show if there any relationship among users in this Network where 1 represent there exist relationship  on the other hand  0 : no relationship